src/dune_rules/lib_rules.ml

```diff
@@ -367,12 +368,23 @@ let setup_build_archives (lib : Dune_file.Library.t) ~top_sorted_modules ~cctx
       ~top_sorted_modules ()
   in
   let* () =
-    Mode.Dict.Set.iter_concurrently modes ~f:(fun mode ->
+    Mode.Dict.Set.iter_concurrently modes.ocaml ~f:(fun mode ->
         build_lib lib ~native_archives ~dir ~sctx ~expander ~flags ~mode ~scope
           ~cm_files)
+  and* () =
+    Memo.when_ modes.melange (fun () ->
+        (* Create empty library target, so that melange libraries modules rules get added to build implicitly *)
```

Comment:
I'm not sure why you need this.

Reply:
If I don't add this rule, then no `ml` sources are added as targets of melange libraries, so the rules set remains empty. Is there a way to somehow add a rule that has `obj_deps` as deps but performs no action?

If I try to remove the `write_file` action:

```ml
let obj_deps =
          Action_builder.paths (Cm_files.melange_objects_and_cms cm_files)
        in
        Super_context.add_rule ~dir sctx ~loc:lib.buildable.loc
          (let open Action_builder.With_targets.O in
          Action_builder.with_no_targets obj_deps))
```

I get this error:
```
This expression has type unit Action_builder.With_targets.t
but an expression was expected of type
  Action.Full.t Action_builder.With_targets.t
Type unit is not compatible with type Action.Full.t
```

Reply:
I’m still not certain what’s the root of the issue. build_lib is for building library archives, which melange doesn’t have. So is it just that you need some fake target to build a library with melange?

For example:

(library
 (name foo))

And you don’t want to write:

dune build foo.cmj

Reply:
What you suggests works, and I can use `.js` instead of `.cmj`. But I have to provide the full path to the `_build` location where the artifact will be placed, e.g:

```
dune build src/.main.objs/melange/main.js
```

works fine.

But if I use the "in-source" path, it won't work:

```
dune build src/main.js
Error: Don't know how to build src/main.js
```

The first option would work if having this hack for .cma is not possible, although it's a bit cumbersome from user perspective.

Reply:
We can work around with build path for now, removed cma hack in https://github.com/jchavarri/dune/pull/4/commits/7882f6ff3b30aea09008c208a4a3f67d33bb2db8.